docs/dep/dep-006-new-go-sdk.md

```diff
@@ -0,0 +1,474 @@
+
+```go
+package main
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"time"
+
+	"github.com/ocurity/dracon/sdk/component"
+	"github.com/ocurity/dracon/api/gen/com/github/ocsf/ocsf-schema/v1"
+)
+
+// simpleComponent parses a Vulnerability Findings report from JSON
+// enriches it and stores the processed output to SQLite.
+type simpleComponent struct{}
+
+func (s simpleComponent) Read(ctx context.Context) (*v1.VulnerabilityFinding, error) {
+	b, err := os.ReadFile("input.json")
+	if err != nil {
+		return nil, fmt.Errorf("failed to read input: %w", err)
+	}
+
+	// Parse to *v1.VulnerabilityFinding
+
+	return &v1.VulnerabilityFinding{}, nil
+}
+
+func (s simpleComponent) Write(ctx context.Context, out *v1.VulnerabilityFinding) error {
+	// Write finding to SQLite
+	return nil
+}
+
+func (s simpleComponent) Transform(ctx context.Context, in *v1.VulnerabilityFinding) (output *v1.VulnerabilityFinding, err error) {
+	// Enrich and transform Input in Output
+	return &v1.VulnerabilityFinding{}, nil
+}
+
+func main() {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	cfg, err := component.NewConfig()
+	if err != nil {
+		log.Fatalf("could not create configuration: %v", err)
+	}
+
+	c, err := component.New(cfg)
+	if err != nil {
+		log.Fatalf("could not create component: %v", err)
+	}
+
+	if err := c.Run(ctx, &simpleComponent{}); err != nil {
+		log.Fatalf("could not run component: %v", err)
+	}
+}
+```
+
+</details>
+
+<a id="proposal-components-catalog"></a>
+
+### Components catalog
+
+Having this simple, transparent and solid foundation for creating new components means that we can
+now provide more out-of-the box ways to get re-usable components and bits.
+
+We can create a separate directory where we can contribute new:
+
+* readers
+* writers
+* transformers
+* components
+
+Each one of these components and actions will have their own go module and dependencies
+to minimise pollution and making individual maintenance easier.
+
+For instance, we can provide re-usable readers, writers and transformers to read/write/transform
+data from/to different source/destinations. Then, such bits can be re-used to create components.
+
+For example:
+
+```
+dracon/
+├── action/
+│   ├── reader/
+│   │   ├── sqlite/
+│   │   │   └── go.mod
+│   │   ├── remote-rpc/
+│   │   │   └── go.mod
+│   │   └── elasticsearch/
+│   │       └── go.mod
+│   ├── writer/
+│   │   ├── sqlite/
+│   │   │   └── go.mod
+│   │   ├── remote-rpc/
+│   │   │   └── go.mod
+│   │   └── elasticsearch/
+│   │       └── go.mod
+│   └── transformer/
+│       ├── atom-enricher/
+│       │   └── go.mod
+│       ├── tagger/
+│       │   └── go.mod
+│       └── prioritiser/
+│           └── go.mod
+```
+
+The components will follow the known structure and just re-use actions or define their own
+to implement the goals of the component.
+
+One can see a scenario where an SQLite reader/writer can be used between multiple components
```

Comment:
great idea! i'd suggest we make where we read/write an implementation detail hidden by the SDK, then the components do not need to implement multiple read/writers and instead can focus on deep transformation and integration with the target things

Reply:
Updated